src/frontend/mame/ui/viewgfx.cpp

```diff
@@ -2,1605 +2,1478 @@
-					{
-						// keep the pointer position and type so we can show touch locations after release
-						m_current_pointer = -1;
-						m_pointer_buttons = 0U;
-						m_pointer_inside = target.map_point_container(
-								event.pointer_x,
-								event.pointer_y,
-								container,
-								m_pointer_x,
-								m_pointer_y);
-					}
-				}
-				break;
-
-			// ignore anything that isn't pointer-related
-			default:
-				break;
-			}
-		}
-
-		// always draw non-touch pointer
-		mame_ui_manager::display_pointer pointers[1]{ { m_machine.render().ui_target(), m_pointer_type, m_pointer_x, m_pointer_y } };
-		if (m_pointer_inside && (0 <= m_current_pointer) && (ui_event::pointer::TOUCH != m_pointer_type))
-			mui.set_pointers(std::begin(pointers), std::end(pointers));
-		else
-			mui.set_pointers(std::begin(pointers), std::begin(pointers));
-
-		// try to display the selected view
-		while (true)
+		m_mode = view((int(m_mode) + 1) % 3);
+		if (0 > m_current_pointer)
 		{
-			switch (m_mode)
-			{
-			case view::PALETTE:
-				if (m_palette.interface())
-					return handle_palette(mui, container, uistate);
-				m_mode = view::GFXSET;
-				break;
-
-			case view::GFXSET:
-				if (m_gfxset.has_gfx())
-					return handle_gfxset(mui, container, uistate);
-				m_mode = view::TILEMAP;
-				break;
-
-			case view::TILEMAP:
-				if (m_machine.tilemap().count())
-					return handle_tilemap(mui, container, uistate);
-				m_mode = view::PALETTE;
-				break;
-			}
+			m_pointer_type = ui_event::pointer::UNKNOWN;
+			m_pointer_x = -1.0F;
+			m_pointer_x = -1.0F;
+			m_pointer_inside = false;
 		}
+		m_bitmap_dirty = true;
 	}
 
-private:
-	enum class view
+	// pause does what you'd expect
+	if (input.pressed(IPT_UI_PAUSE))
 	{
-		PALETTE = 0,
-		GFXSET,
-		TILEMAP
-	};
+		if (m_machine.paused())
+			m_machine.resume();
+		else
+			m_machine.pause();
+	}
 
-	class palette
-	{
-	public:
-		enum class subset
-		{
-			PENS,
-			INDIRECT
-		};
+	// cancel or graphics viewer dismisses the viewer
+	if (input.pressed(IPT_UI_BACK) || input.pressed(IPT_UI_SHOW_GFX))
+		return cancel(uistate);
 
-		palette(running_machine &machine) :
-			m_count(palette_interface_enumerator(machine.root_device()).count())
-		{
-			if (m_count)
-				set_device(machine);
-		}
+	return uistate;
+}
 
-		device_palette_interface *interface() const noexcept
-		{
-			return m_interface;
-		}
+uint32_t gfx_viewer::cancel(bool uistate)
+{
+	if (!uistate)
+		m_machine.resume();
+	m_machine.ui_input().reset();
+	m_current_pointer = -1;
+	m_pointer_type = ui_event::pointer::UNKNOWN;
+	m_pointer_buttons = 0U;
+	m_pointer_x = -1.0F;
+	m_pointer_y = -1.0F;
+	m_pointer_inside = false;
+	m_bitmap_dirty = true;
+	return mame_ui_manager::HANDLER_CANCEL;
+}
 
-		bool indirect() const noexcept
-		{
-			return subset::INDIRECT == m_which;
-		}
+uint32_t gfx_viewer::handle_palette(mame_ui_manager& mui, render_container& container, bool uistate)
```

Comment:
You changed the position on & and * w.r.t the type and variable name everywhere.  People could write books about where it makes sense, but at least mame is almost consistently putting & and * on the variable, so please don't change that, and please do it that way in new code (e.g. gfx_writer.cpp).  Yes, we know Mame is not perfectly consistent everywhere (there are two main styles for a start, K&R and Allman), and we will not do gratuitous code churn to align things everywhere, but in that area we almost are.